### PR TITLE
[full-ci] Implement backchannel logout

### PR DESCRIPTION
**Pull request by dragonchaser**
This PR adds an endpoint for the backchannel logout `/backchannel_logout`.
It can be triggered via a POST request with a special crafted logout token see: [https://openid.net/specs/openid-connect-backchannel-1_0.html](https://openid.net/specs/openid-connect-backchannel-1_0.html) section 2.4 for the structure.

*NOTE:* we have added a special http-handler wrapper to the proxy to intercept predefined routes in the proxy before they are passed on to the services! [https://github.com/dragonchaser/ocis/blob/backchannellogout/services/proxy/pkg/command/server.go#L200](https://github.com/dragonchaser/ocis/blob/backchannellogout/services/proxy/pkg/command/server.go#L200). In future this should be used for terminating stuff like oidc-well-knowns to avoid implementation as middlewares.

refs #3521

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>localApiTests-apiContract-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/21125/50/1) failed. Further test are cancelled...

File: ocis-pkg/oidc/client.go [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+	JWKS        *keyfunc.JWKS
+	jwksLock    *sync.Mutex
+
+	httpClient *http.Client
+}
+
+// supportedAlgorithms is a list of algorithms explicitly supported by this
+// package. If a provider supports other algorithms, such as HS256 or none,
+// those values won't be passed to the IDTokenVerifier.
+var supportedAlgorithms = map[string]bool{
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:72]: Please prefix non exported global vars with `_`. Like `_supportedAlgorithms`

File: ocis-pkg/oidc/client.go [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+	ES256: true,
+	ES384: true,
+	ES512: true,
+	PS256: true,
+	PS384: true,
+	PS512: true,
+}
+
+// NewOIDCClient returns an OIDC client for the given issuer
+func NewOIDCClient(opts ...Option) OIDCClient {
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:85]: What happened to `"Accept interfaces, return structs"`? :wink: This should return a `oidcClient` instance

File: ocis-pkg/oidc/client.go [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+			c.Logger.Error().Err(err).Msg("Failed to create JWKS from resource at the given URL.")
+			return nil
+		}
+	}
+	return c.JWKS
+}
+
+type stringAsBool bool
+
+func (sb *stringAsBool) UnmarshalJSON(b []byte) error {
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:177]: exported func `UnmarshalJSON` should have comment or be unexported

File: ocis-pkg/oidc/client.go [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+	switch string(b) {
+	case "true", `"true"`:
+		*sb = true
+	case "false", `"false"`:
+		*sb = false
+	default:
+		return errors.New("invalid value for boolean")
+	}
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:185]: convert it directly using `strconv.ParseBool` ? https://golang.google.cn/pkg/strconv/#ParseBool

File: ocis-pkg/oidc/client.go [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+// Claims unmarshals the raw JSON object claims into the provided object.
+func (u *UserInfo) Claims(v interface{}) error {
+	if u.claims == nil {
+		return errors.New("oidc: claims not set")
+	}
+	return json.Unmarshal(u.claims, v)
+}
+
+func (c *oidcClient) UserInfo(ctx context.Context, tokenSource oauth2.TokenSource) (*UserInfo, error) {
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:217]: exported func `UserInfo` should have comment or be unexported

File: ocis-pkg/oidc/client.go [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+	if err := c.lookupWellKnownOpenidConfiguration(ctx); err != nil {
+		return nil, err
+	}
+
+	if c.provider.UserinfoEndpoint == "" {
+		if c.provider.UserinfoEndpoint == "" {
+			return nil, errors.New("oidc: user info endpoint is not supported by this provider")
+		}
+	}
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:226]: Isn't that one `if` too much?

File: ocis-pkg/oidc/client.go [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	ct := resp.Header.Get("Content-Type")
+	mediaType, _, parseErr := mime.ParseMediaType(ct)
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:253]: Why name it `parseErr` and not `err`? Imo all errors should be called `err`

File: ocis-pkg/oidc/client.go [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+	}
+	//5. Verify that the Logout Token contains an events Claim whose value is JSON object containing the member name http://schemas.openid.net/event/backchannel-logout.
+	if token.Events.Event == nil {
+		return nil, fmt.Errorf("oidc: logout token must contain logout event")
+	}
+	//6. Verify that the Logout Token does not contain a nonce Claim.
+	type nonce struct {
+		Nonce *string `json:"nonce"`
+	}
+	var n nonce
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:354]:
```suggestion
	var n struct {
		Nonce *string `json:"nonce"`
	}
```
Afais you don't need it afterwards

File: ocis-pkg/oidc/client.go [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+			}
+		} else {
+			return nil, fmt.Errorf("oidc: invalid configuration, clientID must be provided or SkipClientIDCheck must be set")
+		}
+	}
+
+	switch len(jws.Signatures) {
+	case 0:
+		return nil, fmt.Errorf("oidc: logout token not signed")
+	case 1:
```
Comment (kobergj) [inline ocis-pkg/oidc/client.go:380]: Maybe add a comment to highlight the empty case? It's hard to spot.

File: ocis-pkg/oidc/metadata.go [hunk 10]
```diff
@@ -53,6 +54,43 @@ type ProviderMetadata struct {
 	//claim_types_supported
 }
 
+// Logout Token
+type LogoutToken struct {
```
Comment (kobergj) [inline ocis-pkg/oidc/metadata.go:58]: comment of exported type `LogoutToken` should be of form `LogoutToken ...` (with an optional leading article)

File: ocis-pkg/oidc/options.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+type Option func(o *Options)
+
+// Options defines the available options for this package.
+type Options struct {
+	// HTTPClient to use for requests
+	HTTPClient *http.Client
+	// Logger to use for logging, must be set
+	Logger log.Logger
+	// The OpenID Connect Issuer URL
+	OidcIssuer string
```
Comment (kobergj) [inline ocis-pkg/oidc/options.go:22]:
```suggestion
	OIDCIssuer string
```
Abbreviations should be all lower. See `JWKSOptions`

File: ocis-pkg/oidc/metadata.go [hunk 12]
```diff
@@ -53,6 +54,43 @@ type ProviderMetadata struct {
...
+	// The Session Id
+	SessionId string `json:"sid"`
+
+	Events LogoutEvent `json:"events"`
+
+	// Jwt Id
+	JwtID string `json:"jti"`
+}
+
+type LogoutEvent struct {
```
Comment (kobergj) [inline ocis-pkg/oidc/metadata.go:90]: Exported type `LogoutEvent` should have comment or be unexported

File: ocis-pkg/oidc/options.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+	for _, o := range opts {
+		o(&opt)
+	}
+
+	return opt
+}
+
+// WithLogger provides a function to set the openid connect issuer option.
+func WithOidcIssuer(val string) Option {
```
Comment (kobergj) [inline ocis-pkg/oidc/options.go:54]: Comment of func `WithOIDCIssuer` should be of form `WithOIDCIssuer ...` (with an optional leading article)

Note: `OIDC` all upper case as it is an abbreviation

File: ocis-pkg/oidc/options.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+// WithSkipClientIDCheck provides a function to set the skipClientIDCheck option.
+func WithSkipClientIDCheck(val bool) Option {
+	return func(o *Options) {
+		o.SkipClientIDCheck = val
+	}
+}
+
+// WithSkipClientIDCheck provides a function to set the skipClientIDCheck option.
+func WithConfig(val *gOidc.Config) Option {
```
Comment (kobergj) [inline ocis-pkg/oidc/options.go:110]: Comment of exported func `WithConfig` should be of form `WithConfig ...` (with an optional leading article)

File: services/proxy/pkg/command/server.go [hunk 15]
```diff
@@ -36,9 +39,17 @@ import (
 	"github.com/owncloud/ocis/v2/services/proxy/pkg/userroles"
 	"github.com/urfave/cli/v2"
 	microstore "go-micro.dev/v4/store"
-	"golang.org/x/oauth2"
 )
 
+type StaticRouteHandler struct {
```
Comment (kobergj) [inline services/proxy/pkg/command/server.go:44]: Exported func `StaticRouteHandler` should have comment or be unexported

File: services/proxy/pkg/command/server.go [hunk 16]
```diff
@@ -137,7 +188,71 @@ func Server(cfg *config.Config) *cli.Command {
 	}
 }
 
-func loadMiddlewares(ctx context.Context, logger log.Logger, cfg *config.Config) alice.Chain {
+func (h *StaticRouteHandler) handler() http.Handler {
```
Comment (kobergj) [inline services/proxy/pkg/command/server.go:191]: Please move definition of `StaticRouteHandler` here so it and its methods are in one place.

File: services/proxy/pkg/config/config.go [hunk 17]
```diff
@@ -108,6 +108,8 @@ type OIDC struct {
 	UserinfoCache           *Cache `yaml:"user_info_cache"`
 	JWKS                    JWKS   `yaml:"jwks"`
 	RewriteWellKnown        bool   `yaml:"rewrite_well_known" env:"PROXY_OIDC_REWRITE_WELLKNOWN" desc:"Enables rewriting the /.well-known/openid-configuration to the configured OIDC issuer. Needed by the Desktop Client, Android Client and iOS Client to discover the OIDC provider."`
+	ClientID                string `yaml:"client_id" env:"OCIS_OIDC_CLIENT_ID;PROXY_OIDC_CLIENT_ID" desc:"OIDC client ID, which ownCloud Web uses. This client needs to be set up in your IDP."`
+	SkipClientIDCheck       bool   `yaml:"skip_client_id_check" env:"PROXY_OIDC_SKIP_CLIENT_ID_CHECK" desc:"If true will skip checking the configured client id is present in audience claims."`
```
Comment (kobergj) [inline services/proxy/pkg/config/config.go:112]: @mmattel could you do your magic with these strings?

File: services/proxy/pkg/middleware/oidc_auth.go [hunk 18]
```diff
@@ -120,51 +104,25 @@ func (m *OIDCAuthenticator) getClaims(token string, req *http.Request) (map[stri
...
+				// reuse user cache for session id lookup
+				err = m.userInfoCache.Write(&store.Record{
+					Key:    fmt.Sprintf("%s", sid),
+					Value:  []byte(encodedHash),
+					Expiry: time.Until(expiration),
+				})
+			}
+			if err != nil {
+				m.Logger.Error().Err(err).Msg("failed to write session lookup cache")
+			}
```
Comment (kobergj) [inline services/proxy/pkg/middleware/oidc_auth.go:118]: This `err` check should be inside the `if`? Otherwise the `err` is checked twice in `!ok` case

**Review by kobergj — CHANGES_REQUESTED** [review]
Found some smaller ones :wink:

File: services/proxy/pkg/command/server.go [hunk 19]
```diff
@@ -137,7 +188,71 @@ func Server(cfg *config.Config) *cli.Command {
...
+	m := chi.NewMux()
+	var methods = []string{"PROPFIND", "DELETE", "PROPPATCH", "MKCOL", "COPY", "MOVE", "LOCK", "UNLOCK", "REPORT"}
+	for _, k := range methods {
+		chi.RegisterMethod(k)
```
Comment (rhafer) [inline services/proxy/pkg/command/server.go:195]: Hm, this rings a bell. chi.RegisterMethod(k) is racey as it writes to a global map, which other go-routines might be accessing.

We added workarounds for this in other places. See e.g.: https://github.com/cs3org/reva/pull/2785/files and https://github.com/owncloud/ocis/pull/3662

**Review by rhafer — CHANGES_REQUESTED** [review]

File: services/proxy/pkg/config/config.go (hunk 17 above)
Comment (mmattel) [reply services/proxy/pkg/config/config.go:112]: Questions:
* `PROXY_OIDC_CLIENT_ID` --> This client needs to be set up in your IDP.
   - What is necessary to be done in a foreign IDP?
   - As we have an embedded IDP, what do we need to do there?
   - Where does one get the (our) web client ID?
* `PROXY_OIDC_SKIP_CLIENT_ID_CHECK`  --> ... client id is present in audience claims.
Is `audience claim` a well known term?

**Review by kobergj — APPROVED** [review]

**Review by rhafer — APPROVED** [review]

**Pull request merged**